packages/server-wallet/src/onchain-service/store.ts

```diff
@@ -0,0 +1,183 @@
+  > = new Map();
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  constructor() {}
+
+  // Used to save a transaction *before* it is sent to mempool
+  saveTransactionRequest(channelId: Bytes32, tx: NoncedMinimalTransaction): Promise<void> {
+    const existing = this.transactions.get(channelId) || [];
+    const idx = existing.findIndex(
+      e => e.status === 'pending' && (e.data as NoncedMinimalTransaction).nonce === tx.nonce
+    );
+    return Promise.resolve(this._updateTransactions(channelId, idx, {status: 'pending', data: tx}));
+  }
+
+  // Used to save a transaction *after* it has been sent to mempool
+  saveTransactionResponse(channelId: Bytes32, tx: providers.TransactionResponse): Promise<void> {
+    // Find the pending transaction
+    const existing = this.transactions.get(channelId) || [];
+    const idx = existing.findIndex(
+      e => e.status === 'pending' && (e.data as NoncedMinimalTransaction).nonce === tx.nonce
+    );
+    return Promise.resolve(
+      this._updateTransactions(channelId, idx, {status: 'submitted', data: tx})
+    );
+  }
+
+  // Used to save a transaction *after* it is mined
+  saveTransactionReceipt(channelId: Bytes32, receipt: providers.TransactionReceipt): Promise<void> {
+    // Find the submitted transaction
+    const existing = this.transactions.get(channelId) || [];
+    const idx = existing.findIndex(
+      e =>
+        e.status === 'submitted' &&
+        (e.data as providers.TransactionResponse).hash === receipt.transactionHash
+    );
+    return Promise.resolve(
+      this._updateTransactions(channelId, idx, {status: 'success', data: receipt})
+    );
+  }
+
+  // Used to save a transaction if it fails at any point
+  saveFailedTransaction(
+    channelId: Bytes32,
+    tx: NoncedMinimalTransaction,
+    reason: string
+  ): Promise<void> {
+    // Find the submitted transaction
+    const existing = this.transactions.get(channelId) || [];
+    // The transaction can fail at any time before it is sent or during
+    // mining, so look by nonce
+    const idx = existing.findIndex(
+      e => (e.data as providers.TransactionResponse | NoncedMinimalTransaction).nonce === tx.nonce
+    );
+    return Promise.resolve(
+      this._updateTransactions(channelId, idx, {status: 'failed', data: {...tx, reason}})
+    );
+  }
+
+  private _updateTransactions<T extends TransactionStatus>(
+    channelId: Bytes32,
+    previousRecordIndex: number,
+    data: TransactionModel<T>
+  ): void {
+    // Get existing channel transactions
+    const existing = this.transactions.get(channelId) || [];
+
+    // Update record
+    if (previousRecordIndex === -1) {
+      // No record of pending transaction found
+      existing.push(data);
+    } else {
+      // Update existing record
+      existing[previousRecordIndex] = data;
+    }
+    this.transactions.set(channelId, existing);
+  }
+}
+
+export class OnchainServiceStore implements OnchainServiceStoreInterface {
```

Comment:
Maybe we should call this `ChainMonitor` and the other one `TransactionSender` and completely decouple them? They don't need to share any state, I don't think.

Reply:
Depends on how they want to be managed, do you now need to have both a chain monitor and transaction sender within the app/somewhere the wallet can access?

But overall, don't have a problem with decoupling these, especially in v0 when nailing down the wallet <--> chain service flow is more of a v1 scope